### ISSUE TITLE
Pair v1 Polyamorous misdirected messages

### ISSUE THREAD

**Issue by jake-ruyi** [nanomsg/nng#862]
## NNG & Platform details.

HEAD.  macOS.

## Expected Behavior

Based off [docs](https://nanomsg.github.io/nng/man/v1.1.0/nng_pair.7) and [tests](https://github.com/nanomsg/nng/blob/d6bb25e1e0a25cb5aa781ac4f90b513fd5624f50/tests/pair1.c#L196), if:
- `0` is a pair v1 with POLY and listening
- `1a` and `1b` are separate aio using another pair v1 and dialing
Then:
1. `1a` sends nng_msg `m1a0` to `0`
1. `0` receives `m1a0` and calls `nng_msg_dup` on it to make `m01a`
1. `0` calls `nng_msg_get_pipe(m1a0)` and `nng_msg_set_pipe(m01a)` and sends it
1. `1b` simultaneously does the same with `m1b0` (resulting in `m01b`)
1. `1a` receives `m01a` and `1b` receives `m01b`

Basically, calling `nng_msg_set_pipe` and `nng_send_aio` will deliver it back to where it came from.

## Actual Behavior

Sometimes they switch.  Meaning `1a` receives `m01b` and `1b` receives `m01a`.

**Comment by jake-ruyi** [comment]
I originally ran into this with Rust, but my (dubious) C example is similar.
[main.c.txt](https://github.com/nanomsg/nng/files/2772393/main.c.txt)

**Comment by gdamore** [comment]
I suspect that the problem here may be handling of duplicated messages. I'll see if I can figure this out.

**Comment by jeikabu** [comment]
Thought I'd give that a try since it's an easy fix.  Switched `nng_msg_dup` for allocating a new message and it does the same.  I'll try it back in my original test in case there's a problem with the C repro I put together.
Updated: 
[main.c.no_dup.txt](https://github.com/nanomsg/nng/files/2777492/main.c.no_dup.txt)

**Comment by gdamore** [comment]
You wouldn't happen to have a C based reproduction of this would you?

**Comment by jeikabu** [comment]
> [main.c.no_dup.txt](https://github.com/nanomsg/nng/files/2777492/main.c.no_dup.txt)

It's attached

**Comment by jake-ruyi** [comment]
> [main.c.no_dup.txt](https://github.com/nanomsg/nng/files/2777492/main.c.no_dup.txt)

It's attached.

**Comment by jake-ruyi** [comment]
Here's a simpler version:
```c
#include <assert.h>
#include <nng/nng.h>
#include <nng/protocol/pair1/pair.h>
#include <pthread.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

const char *URL = "inproc://test";

nng_socket      pair0_socket = NNG_SOCKET_INITIALIZER;
nng_socket      pair1_socket = NNG_SOCKET_INITIALIZER;
nng_aio *       pair0_aio = NULL;
nng_aio *       pair1a_aio          = NULL;
nng_aio *       pair1b_aio          = NULL;
pthread_mutex_t pair0_mutex         = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t pair1a_mutex        = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t pair1b_mutex        = PTHREAD_MUTEX_INITIALIZER;
bool            is_pair0_receiving  = true;
bool            is_pair1a_receiving = false;
bool            is_pair1b_receiving = false;

#define ASSERT0(x) do { int res = x; assert(res == 0); } while(0)

nng_msg*
create_msg(uint32_t is1a)
{
	nng_msg *msg = NULL;
	ASSERT0(nng_msg_alloc(&msg, 0));
	ASSERT0(nng_msg_append_u32(msg, is1a));
	return msg;
}

void
pair0_callback(void* _unused)
{
	pthread_mutex_lock(&pair0_mutex);
	ASSERT0(nng_aio_result(pair0_aio));
	if (is_pair0_receiving) {
		nng_msg *msg = nng_aio_get_msg(pair0_aio);
		assert(msg != NULL);
		nng_pipe pipe = nng_msg_get_pipe(msg);
		assert(nng_pipe_id(pipe) != -1);
		uint32_t value;
		ASSERT0(nng_msg_trim_u32(msg, &value));

		nng_msg *reply = create_msg(value);
		nng_msg_set_pipe(reply, pipe);
		nng_aio_set_msg(pair0_aio, reply);
		nng_send_aio(pair0_socket, pair0_aio);

		nng_msg_free(msg);
	} else {
		nng_recv_aio(pair0_socket, pair0_aio);
	}
	is_pair0_receiving = !is_pair0_receiving;
	pthread_mutex_unlock(&pair0_mutex);
}

void
pair1_recv(nng_aio* aio, bool is1a)
{
	nng_msg *msg = nng_aio_get_msg(aio);
	assert(msg != NULL);
	uint32_t value;
	ASSERT0(nng_msg_trim_u32(msg, &value));
	assert(value == is1a);
	nng_msg_free(msg);
}

void
pair1a_callback(void* _unused)
{
	pthread_mutex_lock(&pair1a_mutex);
	ASSERT0(nng_aio_result(pair1a_aio));
	if (is_pair1a_receiving) {
		pair1_recv(pair1a_aio, true);
		is_pair1a_receiving = !is_pair1a_receiving;
		nng_msg* msg = create_msg(true);
		nng_aio_set_msg(pair1a_aio, msg);
		nng_send_aio(pair1_socket, pair1a_aio);
	} else {
		is_pair1a_receiving = !is_pair1a_receiving;
		nng_recv_aio(pair1_socket, pair1a_aio);
	}
	pthread_mutex_unlock(&pair1a_mutex);
}

void
pair1b_callback(void* _unused)
{
	pthread_mutex_lock(&pair1b_mutex);
	ASSERT0(nng_aio_result(pair1b_aio));
	if (is_pair1b_receiving) {
		pair1_recv(pair1b_aio, false);
		is_pair1b_receiving = !is_pair1b_receiving;
		nng_msg* msg = create_msg(false);
		nng_aio_set_msg(pair1b_aio, msg);
		nng_send_aio(pair1_socket, pair1b_aio);
	} else {
		is_pair1b_receiving = !is_pair1b_receiving;
		nng_recv_aio(pair1_socket, pair1b_aio);
	}
	pthread_mutex_unlock(&pair1b_mutex);
}

int
main(int argc, char **argv)
{
	nng_msg* msg;

	// Listener "0"
	ASSERT0(nng_pair1_open(&pair0_socket));
	ASSERT0(nng_setopt_bool(pair0_socket, NNG_OPT_PAIR1_POLY, true));
	ASSERT0(nng_listen(pair0_socket, URL, NULL, 0));
	ASSERT0(nng_aio_alloc(&pair0_aio, &pair0_callback, NULL));
	nng_recv_aio(pair0_socket, pair0_aio);

	usleep(50000);
	ASSERT0(nng_pair1_open(&pair1_socket));

	// Dialer "1a"
	ASSERT0(nng_dial(pair1_socket, URL, NULL, 0));
	ASSERT0(nng_aio_alloc(&pair1a_aio, &pair1a_callback, NULL));
	msg = create_msg(true);
	nng_aio_set_msg(pair1a_aio, msg);
	nng_send_aio(pair1_socket, pair1a_aio);
	
	// Dialer "1b"
	ASSERT0(nng_dial(pair1_socket, URL, NULL, 0));
	ASSERT0(nng_aio_alloc(&pair1b_aio, &pair1b_callback, NULL));
	msg = create_msg(false);
	nng_aio_set_msg(pair1b_aio, msg);
	nng_send_aio(pair1_socket, pair1b_aio);

	while (1) {
		usleep(10000);
	}
	return 0;
}
```

**Comment by gdamore** [comment]
Thanks.  I will definitely investigate this.  I'm in the middle of another round of changes, that I want to finish up first.

**Comment by gdamore** [comment]
I have an idea on fixing this.. stay tuned... I've not forgotten.  (I think nni_msg_dup(), which is used internally, has a bug here, in that it does not propogate the pipe id.)

**Comment by jake-ruyi** [comment]
It's no rush.  Just found it when trying to write various tests.

**Comment by qsodev** [comment]
I also ran into this issue.  I can confirm: `nng_msg_dup()` does not copy pipe id correctly.

**Comment by jake-ruyi** [comment]
So, PR#909 fixes `nng_msg_dup()` to copy the pipe correctly.  It doesn't fix this issue, however.

**Comment by qsodev** [comment]
I haven't widely tested but it seems to fix the issue at least for my scenario.

**Comment by gdamore** [comment]
I think the fix in 909 probably does fix the real issue, but I think your test program may be busted too.

The concern here is that with your pair1a vs pair1b, you can wind up conflating results.  Really you need to have three sockets as part of your test. One polyamorous socket, and two separate sockets -- using separate dialers on on the same pipe is not sufficient, because you have no way to specify which dialer's pipe is used on the initial outbound message from socket 1.

I'll refactor the test code and update.

**Comment by gdamore** [comment]
```
#include <assert.h>
#include <nng/nng.h>
#include <nng/protocol/pair1/pair.h>
#include <nng/supplemental/util/platform.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char *URL = "inproc://test";

nng_socket pair0_socket       = NNG_SOCKET_INITIALIZER;
nng_aio *  pair0_aio          = NULL;
bool       is_pair0_receiving = true;

#define ASSERT0(x)                \
	do {                      \
		int res = x;      \
		assert(res == 0); \
	} while (0)

struct partner {
	uint32_t   num;
	nng_socket socket;
	nng_aio *  aio;
	bool       recving;
	uint32_t   nrx;
	uint32_t   ntx;
} * partners;

nng_msg *
create_msg(uint32_t is1a)
{
	nng_msg *msg = NULL;
	ASSERT0(nng_msg_alloc(&msg, 0));
	ASSERT0(nng_msg_append_u32(msg, is1a));
	return (msg);
}

void
pair0_callback(void *_unused)
{
	int rv;
	(void) _unused;

	if ((rv = nng_aio_result(pair0_aio)) == NNG_ECLOSED) {
		return;
	}
	if (rv != 0) {
		printf("pair0: %s\n", nng_strerror(rv));
	}
	assert(rv == 0);
	if (is_pair0_receiving) {
		is_pair0_receiving = false;

		nng_msg *msg = nng_aio_get_msg(pair0_aio);
		assert(msg != NULL);
		nng_pipe pipe = nng_msg_get_pipe(msg);
		assert(nng_pipe_id(pipe) != -1);
		uint32_t value;
		ASSERT0(nng_msg_trim_u32(msg, &value));

		nng_msg *reply = create_msg(value);
		nng_msg_set_pipe(reply, pipe);
		nng_aio_set_msg(pair0_aio, reply);
		nng_send_aio(pair0_socket, pair0_aio);

		nng_msg_free(msg);
	} else {
		is_pair0_receiving = true;
		nng_recv_aio(pair0_socket, pair0_aio);
	}
}

void
partner_callback(void *arg)
{
	struct partner *p   = arg;
	nng_aio *       aio = p->aio;
	nng_msg *       msg;
	int             rv;

	if ((rv = nng_aio_result(aio)) == NNG_ECLOSED) {
		return;
	}
	if (rv != 0) {
		printf("ERROR! %d %s\n", rv, nng_strerror(rv));
	}
	ASSERT0(nng_aio_result(aio));
	if (p->recving) {
		uint32_t value;
		msg = nng_aio_get_msg(aio);
		ASSERT0(nng_msg_trim_u32(msg, &value));
		nng_msg_free(msg);
		assert(value == p->num);
		msg        = create_msg(p->num);
		p->recving = false;
		p->nrx++;
		nng_aio_set_msg(aio, msg);
		nng_send_aio(p->socket, aio);
	} else {
		// Send completed.
		p->recving = true;
		p->ntx++;
		nng_recv_aio(p->socket, aio);
	}
}

int
main(int argc, char **argv)
{
	int nsec      = 10;
	int npartners = 2;
	if (argc > 1) {
		nsec = atoi(argv[1]);
	}
	if (argc > 2) {
		npartners = atoi(argv[2]);
	}

	// Listener "0"
	partners = calloc(npartners, sizeof(struct partner));
	ASSERT0(nng_pair1_open(&pair0_socket));
	ASSERT0(nng_setopt_bool(pair0_socket, NNG_OPT_PAIR1_POLY, true));
	ASSERT0(nng_listen(pair0_socket, URL, NULL, 0));
	ASSERT0(nng_aio_alloc(&pair0_aio, &pair0_callback, NULL));
	nng_recv_aio(pair0_socket, pair0_aio);
	nng_msleep(50);

	for (int i = 0; i < npartners; i++) {
		nng_msg *       msg;
		struct partner *p = &partners[i];
		p->num            = i + 1;
		p->nrx            = 0;
		p->ntx            = 0;
		p->recving        = false;

		ASSERT0(nng_aio_alloc(&p->aio, partner_callback, p));
		ASSERT0(nng_pair1_open(&p->socket));
		ASSERT0(nng_dial(p->socket, URL, NULL, 0));
		msg = create_msg(p->num);
		nng_aio_set_msg(p->aio, msg);
		nng_send_aio(p->socket, p->aio);
	}

	printf("RUNNING FOR %d SECONDS (%d PARTNERS)\n", nsec, npartners);
	nng_msleep(1000 * nsec);
	printf("CLOSING....\n");
	nng_close(pair0_socket);
	for (int i = 0; i < npartners; i++) {
		struct partner *p = &partners[i];
		nng_close(p->socket);
		nng_aio_free(p->aio);
		printf("PARTNER %d: TX %u RX %u\n", p->num, p->ntx, p->nrx);
	}
	return (0);
}
```

Notice that we don't need any locks, and we can adjust the number of seconds and number of partners.  The old code failed right away due to incorrect use of partners.  This new code seems not to fail at all.

**Comment by gdamore** [comment]
I don't think there is actually a misdirection bug here at all.

**Comment by gdamore** [comment]
Submitter, please review the code, and see if you can reproduce.

**Comment by jake-ruyi** [comment]
Ok, so the answer to the original question is: no, separate aio/dialers isn't sufficient, they need their own sockets.  This result is perhaps surprising but not "wrong".
It's been a while since I looked at the docs, but I think I got the impression that "dialers create pipes" and "polyamorous pairs honor them".

**Comment by Stathastic** [comment]
I am also trying to implement communication between one server and multi clients..Is this problem fixed now?
